### PR TITLE
Merchant Items Index

### PR DESCRIPTION
**Pull request by samueldevine**
Pushing this up so other can work with the index page as needed! I'm a little unsure about how to approach the `unit_price` in the create method on the `merchant_items_controller` so if anyone has any feedback there I'd appreciate it! Unit_price should be stored as in integer in cents, but it doesn't really make sense for the user to enter it that way.

### CODE REVIEW

File: app/controllers/merchant_items_controller.rb [hunk 1]
```diff
@@ -1,4 +1,26 @@
...
+    @merchant = Merchant.find(params[:merchant_id])
+    @item = Item.new
+  end
+
+  def create
+    params[:unit_price] = (params[:unit_price].to_f * 100).to_i
+    merchant = Merchant.find(params[:merchant_id])
+    item = merchant.items.create(item_params)
+    redirect_to "/merchants/#{params[:merchant_id]}/items"
+  end
```
Comment (cvalesares) [inline app/controllers/merchant_items_controller.rb:20]: We talked about this already but we might want to revisit this during refactoring to see if there is a better way to do the conversion than in the create method (or maybe just a DRYer way within this method). It gets the job done though  😀

File: app/models/merchant.rb [hunk 2]
```diff
@@ -4,27 +4,40 @@ class Merchant < ApplicationRecord
...
     Customer.joins(invoices: [:transactions, [invoice_items: [item: [:merchant]]]])
       .where(transactions: { result: 'success' })
-      .where(merchants: { id: id })  #second id is the id of the merchant that's getting called out
+      .where(merchants: { id: id })
       .group(:id)
       .order(Arel.sql('COUNT(transactions.id) DESC'))
       .limit(5)
+      .select(
+        "customers.first_name as first_name,
+        customers.last_name as last_name,
```
Comment (cvalesares) [inline app/models/merchant.rb:13]: Would it be beneficial to create a helper method to dry up adding the first and last name strings? I feel like we could use that in a few places throughout the project.

File: app/models/merchant.rb [hunk 3]
```diff
@@ -4,27 +4,40 @@ class Merchant < ApplicationRecord
...
+      .where(transactions: { result: 'success' })
+      .group(:id)
+      .order(Arel.sql('SUM(invoice_items.quantity * invoice_items.unit_price) DESC'))
+      .limit(5)
+      .select(
+        "items.id as item_id,
+        items.name as item_name,
+        SUM(invoice_items.quantity * invoice_items.unit_price / 100) as revenue"
+      )
+  end
```
Comment (cvalesares) [inline app/models/merchant.rb:42]: this thing is a beast! nicely done

File: db/migrate/20211103174459_create_invoices.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@ class CreateInvoices < ActiveRecord::Migration[5.2]
   def change
     create_table :invoices do |t|
       t.references :customer, foreign_key: true
-      t.string :status
+      t.integer :status
```
Comment (cvalesares) [inline db/migrate/20211103174459_create_invoices.rb]: I forget where we landed on this... would it be beneficial to set a default status as "in progress" for these enums? Or do we want to keep it as is? Might make life easier if we need to add more data down the road.

File: spec/features/admin_invoices/index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe 'the admin invoice index' do
+  describe 'as an admin' do
+    before :each do
+      customer = Customer.create!(first_name: 'Taylor', last_name: 'Swift')
+      @invoice_1 = customer.invoices.create!(status: 'in progress')
+      @invoice_2 = customer.invoices.create(status: 'cancelled')
+      @invoice_3 = customer.invoices.create(status: 'completed')
+    end
```
Comment (cvalesares) [inline spec/features/admin_invoices/index_spec.rb]: Should these be @customer if they are used within a before :each block?

**Review by cvalesares — CHANGES_REQUESTED** [review]
Great job! There is only 1 small tweak that needs to be made before pushing but most of the comments are general refactoring ideas that we don't need to stress about.

File: app/models/merchant.rb (hunk 2 above)
Comment (samueldevine) [reply app/models/merchant.rb:13]: possibly! I think we could also add them together in the SQL if we wanted to, like `"customers.first_name + ' ' + customers.last_name as full_name"` or something

File: db/migrate/20211103174459_create_invoices.rb (hunk 4 above)
Comment (samueldevine) [reply db/migrate/20211103174459_create_invoices.rb]: good point! totally glossed over the default option, just added them in

File: spec/features/admin_invoices/index_spec.rb (hunk 5 above)
Comment (samueldevine) [reply spec/features/admin_invoices/index_spec.rb]: customer isn't ever being referenced in the tests which is why I didn't make it an instance variable. not sure if that's important for performance or anything but I just assumed the less instance variables, the better 🤷‍♂️

**Pull request merged**